File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.java
class Solution {
    public int minDeletions(String s) {
        int arr[]=new int[26];
        for(int i:arr)
            i=0;
        for (int i=0;i<s.length();i++)
            arr[s.charAt(i)-97]++;
        Arrays.sort(arr);
        /*for(int i:arr)
            System.out.println(i);*/
        int c=0,max=s.length();
        for(int i=25;i>=0;i--)
        {
            if(arr[i]>0)
            {
                if(arr[i]>max)
                {
                    c=c+(arr[i]-max);
                    arr[i]=max;
                }
                if(arr[i]-1>0)
                    max=arr[i]-1;
                else
                    max=0;
            }
        }
        return c;
    }
}

File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.py
class Solution:
    def minDeletions(self, s: str) -> int:
        
        # Store the frequency of each character
        frequency = [0] * 26
        for char in s:
            frequency[ord(char) - ord('a')] += 1
            
        # Add all non-zero frequencies to max priority queue
        # Create a max priority queue by flipping the sign of each element
        pq = [-freq for freq in frequency if freq != 0]
        heapq.heapify(pq)
        
        delete_count = 0
        while len(pq) > 1:
            # Flip the sign back to positive when removing from the max priority queue
            top_element = -heapq.heappop(pq)
            
            # If the top two elements in the priority queue are the same
            if top_element == -pq[0]:
                # Decrement the popped value and push it back into the queue
                if top_element - 1 > 0: 
                    top_element -= 1
                    heapq.heappush(pq, -top_element)

                delete_count += 1
        
        return delete_count